src/Stratis.Bitcoin/P2P/Peer/NetworkPeer.cs

```diff
@@ -1048,10 +1162,24 @@ public InventoryType AddSupportedOptions(InventoryType inventoryType)
             return inventoryType;
         }
 
+        /// <summary>
+        /// Disconnects the peer and cleans up.
+        /// </summary>
+        /// <param name="reason">Human readable reason for disconnecting.</param>
+        /// <param name="exception">Exception because of which the disconnection happened, or <c>null</c> if there were no exception.</param>
+        public void Dispose(string reason, Exception exception = null)
```

Comment:
I don't think that dispose should have 'reason' or 'exception' as input values. This is not a classic and expected implementation of Dispose. 

I suggest replacing every usage of this dispose with disconnect method and when the object is actually being disposed add checking if peer is connected. So it will be like: 

```
   public void Dispose()
        {
            if (this.Connected)
             this.Disconnect(reason: "Peer was disposed");

        }
```

Reply:
Actually it looks like you already have a Dispose overload that does that so just removing 'Dispose(string reason, Exception exception = null)' would do.

Reply:
Or if you absolutely need to do both things- rename this dispose overload to "DisconnectAndDispose"

Reply:
I was thinking about it and I have not found a perfect solution for this. I at least tried to distinguish where to use which method. 

I agree it is not optimal, I was thinking about alternative version, but then decided for this one. The alternative version was that you would specify (in constructor?) the default disconnection message for the particular peer instance. So then you would just call dispose. But then you wouldn't be able to pass exception.

If you check the code that non-parametric Dispose is doing just what you are asking for, so that is there.

I'd be happy to get more opinions on this as I'm undecided which way to take here.

Reply:
However, I don't hink DisconnectAndDispose is any better than just parametric Dispose.

Reply:
Note that the default Dispose is there and you can (and it is used in tests) use `using` clause for this and it will work as expected.